State: Add primary site selectors

tyxla opened the pull request:
This PR does the following:

* Dispatches a Flux action each time the sites list is fetched.
* Introduces a mechanism to sync Flux actions to Redux (as borrowed from #2167).
* Uses the above mechanism to sync the new Flux action to the Redux one, so we receive the sites in the Redux tree.
* Introduces the `getPrimarySite` selector.
* Includes tests for the new selector and the sync module.

It depends on D3252-code, which adds an additional `is_primary` prop to each site.

To test:
* Checkout this branch.
* Point `public-api.wordpress.com` to your sandbox.
* Apply D3252-code on your sandbox.
* Clear your Redux state.
* Go to a page that does not specifically fetch the sites Redux tree (for example `/me`.
* Verify each site from the `/me/sites` request contains a `is_primary` prop, and it's true only for your primary site. If there's no such property, you're using the production API.
* Verify that once `/me/sites` loads, the sites are stored in the `sites.items` Redux tree.
* Verify that the selector tests pass: `npm run test-client client/state/sites/test/selectors.js -- --grep "getPrimarySite"`
* Verify that the sync module tests pass: `npm run test-client client/lib/redux-flux-sync/`

mtias commented on client/state/sites/selectors.js:
```diff
@@ -73,6 +73,38 @@ export const getSite = createSelector(
+ * @param  {Object}   state  Global state tree
+ * @return {?Number}         The ID of the primary site
+ */
+export function getPrimarySiteId( state ) {
```
Why not return the site?

tyxla replied:
Makes sense. That would also ensure compatibility with `SitesList.getPrimary()`, so I'm changing it now.

mtias commented on client/state/sites/selectors.js:
```diff
@@ -73,6 +73,38 @@ export const getSite = createSelector(
+ * @return {?Number}         The ID of the primary site
+ */
+export function getPrimarySiteId( state ) {
+	const site = find( state.sites.items, 'is_primary' );
```
Where does `is_primary` come from? Before, we had to assign an attribute of `primary` in sites-list. ee previous attempt https://github.com/Automattic/wp-calypso/pull/4819

mtias replied:
Ah, I missed your comment on the description :)

mtias commented on client/state/sites/selectors.js:
```diff
@@ -73,6 +73,38 @@ export const getSite = createSelector(
+ * @param  {Number}  siteId Site ID
+ * @return {?Boolean}       Whether this is the primary site
+ */
+export function isPrimarySiteId( state, siteId ) {
```
Do you have a use case for this selector?

tyxla replied:
Guess not. I'm removing it for now - we can always add it if we need it.

aduth commented:
Currently we only sync the currently selected site to Redux:

https://github.com/Automattic/wp-calypso/blob/c42a7f7/client/my-sites/controller.js#L172

Because of this, there's a good chance that unless I've already visited a page for the primary site, it won't be possible to find it in Redux state (since we don't have all sites data there).

That is, until something like #7775 or equivalent solution exists.

tyxla commented:
@aduth AFAIK, this is true for many other selectors. Isn't that the reasoning behind having query components in the first place? I think we use `<QuerySites allSites />` in that case?

aduth commented:
Yes, but `/me/sites` is an incredibly slow endpoint and we must avoid duplicating requests between `sites-list` and Redux. This is one of our biggest challenges with the "Remove SitesList" project.

tyxla commented:
@aduth, yup, that makes sense. And I guess that endpoint gets even more slow if you have like too many sites?

I'm thinking of using an alternative, although a more time-consuming approach:
1. Reduxifying `lib/user-settings`
2. Using the `primary_site_ID` that comes from `/me/settings` - it's a rather fast endpoint.

Thoughts?

aduth commented:
That would help to get the site ID, yes, but it still wouldn't address the need for having access to the full site object, which could then mislead other developers to thinking they could `getSite( state, getPrimarySiteId( state ) )`

tyxla commented:
After discussing with @aduth in DM, I've taken a different approach. 

Since we're already polling for fetching the sites list on all pages, we don't want to create a duplicate request to do it for the Redux state. So I've changed the sites list to dispatch a Flux action upon fetch, and borrowed the approach from #2167 to sync the Flux action to the Redux one. This way the sites will be stored in the Redux tree as soon as they're available, and we'll be able to use any selectors that need the `state.sites.items` tree.

In addition, I've removed the `isPrimarySiteId` selector (it's unnecessary) and updated the `getPrimarySite` selector to actually return the normalized site object. Reasoning behind this change: this is how the legacy `SitesList.getPrimary` method works.

If we land this, we'll be able to add some more selectors that rely on the presence of `state.sites.items`.

tyxla commented:
/cc @gwwar @aduth @mtias for a review

aduth commented:
@tyxla I should have mentioned that there was a recent (now punted) effort to sync all of sites list to Redux in #7775. The primary issue there is that because of the amount of data we expect to exist in sites (at least for users with multiple sites), the duplication and especially persisted duplication is unacceptable. Furthermore, there is no `SITES_RECEIVE` in Flux. `sites-list` is an [emitter object](https://github.com/Automattic/wp-calypso/blob/96001c7cbd1640ba4f8513faf51f5322dcadd71c/docs/our-approach-to-data.md#first-era-emitter-objects-june-2014---april-2015), with a [few Flux-based compatibility handlers](https://github.com/Automattic/wp-calypso/blob/96001c7cbd1640ba4f8513faf51f5322dcadd71c/client/lib/sites-list/index.js#L17-L42).

tyxla commented:
That makes sense - it would be best if the duplication is live for as short as possible. Punting this for until we get rid of all `sitesList.getSelectedSite()` usages - #8726.

gwwar commented:
One thing we could try is experiment with the redux tree as being the source of truth, and hack SitesList to poll from the redux tree rather than hitting the endpoints directly. This could make porting certain sites list usages much easier.
